### DISCUSSION TITLE
Data mutations - transactions, data validation, etc?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#462] [Q&A] [answered]
I'm curious to know how/if PocketBase attempts to solve data integrity concerns such as locking, transactions, and validation. I see that I can implement basic read/write security rules, but what is stopping a rogue client from writing invalid data, particularly invalid field values and orphaned records? 

Maybe I'm a dinosaur, but I have this same thought about every BaaS including Supabase and Firebase. For example, Firebase's data security rules seem insufficient to describe use cases where multiple fields are supposed to be modified together or not at all, especially as the use cases grown in number and complexity of constraints. 

Would the answer be to prohibit client-side mutations and extend PocketBase with a REST API to perform them instead?

**Comment by ganigeorgiev** [comment, marked as answer]
PocketBase collections requires the developers to always define schema for their collections. Validations of the inserted data are performed based on your defined field types. For example, if you have an `email` field, you'll not be able to insert a non-email formatted string.

> particularly invalid field values and orphaned records?

I'm not sure what do you mean by "orpahned" records. Each `relation` type field has "Cascade" options that you can configure whether for example you want to delete the parent record or set the field to the respective zero-value, aka. empty string or array (PocketBase doesn't have `null` values).

PocketBase has ["Collection API rules"](https://pocketbase.io/docs/manage-collections/#collection-api-rules) to control who can read or modify the related records. Depending on how you use PocketBase, there are usually 2 ways to prevent client users to create/update records:

1. using an API rule and for example checking the submitted data using the `@request.data.*` filter, eg. `@requst.data.name != ""`
2. using the [before* event hooks](https://pocketbase.io/docs/event-hooks/)

---

About the transactions - PocketBase web API currently doesn't support bulk record operations (this will be handled in #48). If you use PocketBase as framework, you can start a transaction and execute multiple queries together using the dao service: [`dao.RunInTransaction(...)`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.7.2/daos#Dao.RunInTransaction).

**Reply by benallfree** [reply]
Thanks for such a fast and thorough answer, and thank you for all the work on PocketBase! I'm writing a PocketBase provider today to see how it fits with my application.

**Reply by ffunenga** [reply]
> If you use PocketBase as framework, you can start a transaction and execute multiple queries together using the dao service: [dao.RunInTransaction(...)](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.7.2/daos#Dao.RunInTransaction).

What does any of this mean, in practical terms? Can you give a simple usage example? For example, how to execute two `INSERT INTO` statements inside a transaction, each statement inserting a new entry into a respective SQL table.

**Reply by ganigeorgiev** [reply]
@ffunenga Here is a mixed read/write transaction using the Dao:

```go
txErr := app.Dao().RunInTransaction(func(txDao *daos.Dao) error {
	collection, err := txDao.FindCollectionByNameOrId("example")
	if err != nil {
		return err
	}

	// Inserts 2 new records in the "example" collection.
	// 
	// NB! Dao doesn't perform validations.
	// It just persist the record in the database.
	// If you need data validations, check `forms.NewRecordUpsert()`
	// -----------------------------------------------------------------
	record1 := models.NewRecord(collection)
	record1.SetDataValue("someField", "test1")
	if err := txDao.SaveRecord(record1); err != nil {
		return err
	}

	record2 := models.NewRecord(collection)
	record2.SetDataValue("someField", "test2")
	if err := txDao.SaveRecord(record2); err != nil {
		return err
	}

	return nil
})
```

**The "Use as framework" documentation will get updated soon with more examples and details, but currently #376 is with highest priority because it blocks all other feature requests.**

**Reply by ffunenga** [reply]
Thanks!
(Note: I am struggling **a lot** with Go.)
